core/modules/wildcard_match.cc

```diff
@@ -397,30 +407,33 @@ CommandResponse WildcardMatch::CommandGetRules(const bess::pb::EmptyArg &) {
     }
     f->set_num_bytes(field.size);
   }
+}
+
+// Sets up default gate, adds a rule for each (gate, priority) entry,
+// and then calls put_field on each (field_mask, field_match, size).
+template <typename rule_t, typename T, typename Func>
+void WildcardMatch::GetRules(T &resp, Func put_field) {
+  resp.set_default_gate(default_gate_);
 
+  // Each tuple provides a single mask, which may have many data-matches.
   for (auto &tuple : tuples_) {
     wm_hkey_t mask = tuple.mask;
+    // Each entry in the hash table has priority, ogate, and the data
+    // (one datum per field, under the mask for this field).
     for (auto &entry : tuple.ht) {
-      bess::pb::WildcardMatchRule *rule = resp.add_rules();
-      rule->set_priority(entry.second.priority);
+      // Create the rule instance
+      rule_t *rule = resp.add_rules();
       rule->set_gate(entry.second.ogate);
-
+      rule->set_priority(entry.second.priority);
       uint8_t *entry_data = reinterpret_cast<uint8_t *>(entry.first.u64_arr);
       uint8_t *entry_mask = reinterpret_cast<uint8_t *>(mask.u64_arr);
+      // Then fill in each field
       for (auto &field : fields_) {
-        uint64_t data = 0;
-        bess::utils::bin_to_uint64(&data, entry_data + field.pos, field.size,
-                                   true);
-        rule->add_values(data);
-        uint64_t mask_data = 0;
-        bess::utils::bin_to_uint64(&mask_data, entry_mask + field.pos,
-                                   field.size, true);
-        rule->add_masks(mask_data);
+        put_field(rule, entry_data + field.pos, entry_mask + field.pos,
```

Comment:
Given that there is only one caller of `GetRules` yet, can we inline the function `put_field` here for now? Template parameter `Func` doesn't seem to do much here.